eip_operator/src/controller/node.rs

```diff
@@ -83,41 +96,57 @@ impl k8s_controller::Context for Context {
             .swap_remove(0);
         let instance_description =
             crate::aws::describe_instance(&self.ec2_client, instance_id).await?;
-
         let eni_id = crate::aws::get_eni_from_private_ip(&instance_description, node_ip)
             .ok_or(Error::NoInterfaceWithThatIp)?;
-        if eip_description.network_interface_id != Some(eni_id.to_owned())
-            || eip_description.private_ip_address != Some(node_ip.to_owned())
-        {
-            match crate::eip::set_status_should_attach(&eip_api, &eip, &eni_id, node_ip, &name)
-                .await
-            {
-                Ok(_) => {
-                    info!("Found matching Eip, attaching it");
-                    let association_id = crate::aws::associate_eip(
-                        &self.ec2_client,
-                        allocation_id,
-                        &eni_id,
-                        node_ip,
-                    )
-                    .await?
-                    .association_id
-                    .ok_or(Error::MissingAssociationId)?;
-                    crate::eip::set_status_association_id(&eip_api, &eip_name, &association_id)
-                        .await?;
-                }
-                Err(Error::Kube {
-                    source: kube::Error::Api(ErrorResponse { reason, .. }),
-                }) if reason == "Conflict" => {
-                    warn!(
-                        "Node {} failed to claim eip {}, rescheduling to try another",
-                        name, eip_name
-                    );
-                    return Ok(Some(Action::requeue(Duration::from_secs(3))));
-                }
-                Err(e) => return Err(e),
-            };
+
+        let network_interface_mismatch =
+            eip_description.network_interface_id != Some(eni_id.to_owned());
+        let private_ip_mismatch = eip_description.private_ip_address != Some(node_ip.to_owned());
+        let attach_result = match (
+            network_interface_mismatch,
+            private_ip_mismatch,
+            node_conditon_ready_status.as_str(),
+        ) {
+            (true, false, "True") | (false, true, "True") | (true, true, "True") => {
+                crate::eip::set_status_should_attach(&eip_api, &eip, &eni_id, node_ip, &name).await
+            }
+            // Node is in a ready state of `unknown`. Disassociate the EIP.
+            (false, false, "Unknown") => {
+                // This code path runs a limited number of times once a node goes unresponsive.
+                // The EIP will be disassociated and then this apply() will exit early since no EIP is associated
+                warn!("Node {} is in Unknown state, disassociating EIP", &name);
+                crate::aws::disassociate_eip(&self.ec2_client, &eip_name).await?;
```

Comment:
if for any reason the `eip-operator` crashed between executing `disassociate_eip` and `set_status_detached`, would we be able to reconcile and clear the status correctly?

Reply:
I like this idea! Leaving for a separate PR as discussed.